solidity/test/OwnableMOfNSet.t.sol

```diff
@@ -0,0 +1,250 @@
+        set.setThreshold(domain, threshold);
+    }
+
+    function testSetThresholds(uint32[] memory domains, address value) public {
+        vm.assume(value != address(0));
+        uint8 threshold = 1;
+        uint8[] memory thresholds = new uint8[](domains.length);
+        // `domains` could contain repeats
+        bool[] memory skip = new bool[](domains.length);
+        for (uint256 i = 0; i < domains.length; i++) {
+            thresholds[i] = threshold;
+            if (set.contains(domains[i], value)) {
+                skip[i] = true;
+            } else {
+                set.add(domains[i], value);
+            }
+        }
+        for (uint256 i = 0; i < domains.length; i++) {
+            if (skip[i]) continue;
+            vm.expectEmit(true, true, false, true, address(set));
+            emit ThresholdSet(domains[i], threshold);
+            bytes32 commitment = keccak256(
+                abi.encodePacked(uint8(threshold), [value])
+            );
+            emit CommitmentUpdated(domains[i], commitment);
+        }
+        set.setThresholds(domains, thresholds);
+    }
+
+    function testContains(uint32 domain, address value) public {
+        vm.assume(value != address(0x0));
+        assertFalse(set.contains(domain, value));
+        set.add(domain, value);
+        assertTrue(set.contains(domain, value));
+        set.remove(domain, value);
+        assertFalse(set.contains(domain, value));
+    }
+
+    function testLength(uint32 domain, address value) public {
+        vm.assume(value != address(0x0));
+        assertEq(set.length(domain), 0);
+        set.add(domain, value);
+        assertEq(set.length(domain), 1);
+    }
+
+    function testThreshold(uint32 domain, address value) public {
+        vm.assume(value != address(0x0));
+        assertEq(set.threshold(domain), 0);
+        set.add(domain, value);
+        set.setThreshold(domain, 1);
+        assertEq(set.threshold(domain), 1);
+    }
+
+    function testValuesAndThreshold(uint32 domain, address value) public {
+        vm.assume(value != address(0x0));
+        (address[] memory values, uint8 threshold) = set.valuesAndThreshold(
+            domain
+        );
+        address[] memory emptySet = new address[](0);
+        assertEq(abi.encodePacked(values), abi.encodePacked(emptySet));
+        assertEq(threshold, 0);
+
+        set.add(domain, value);
+        set.setThreshold(domain, 1);
+        (values, threshold) = set.valuesAndThreshold(domain);
+        address[] memory nonEmptySet = new address[](1);
+        nonEmptySet[0] = value;
+        assertEq(abi.encodePacked(values), abi.encodePacked(nonEmptySet));
+        assertEq(threshold, 1);
+    }
+
+    function testSetMatches(
+        uint32 domain,
+        address firstValue,
+        uint8 numValues,
+        uint8 threshold
+    ) public {
+        vm.assume(0 < threshold && threshold <= numValues);
+        vm.assume(numValues < uint160(firstValue) && numValues < 32);
```

Comment:
what does numValues < firstValue do

Reply:
Avoid underflow b/c we subtract numValues from firstValue